jina/drivers/multimodal.py

```diff
@@ -0,0 +1,114 @@
+__copyright__ = "Copyright (c) 2020 Jina AI Limited. All rights reserved."
+__license__ = "Apache-2.0"
+
+from collections import defaultdict
+from typing import Iterable, Tuple, Dict
+
+import numpy as np
+
+from .encode import BaseEncodeDriver
+from .helper import pb2array, array2pb
+from ..proto import jina_pb2
+
+
+def _extract_doc_content(doc: 'jina_pb2.Document'):
+    """Returns the content of the document with the following priority:
+    If the document has an embedding, return it, otherwise return its content.
+    """
+    if doc.embedding.buffer:
+        return pb2array(doc.embedding)
+    else:
+        return doc.text or doc.buffer or (doc.blob and pb2array(doc.blob))
+
+
+def _extract_modalities_from_document(doc: 'jina_pb2.Document'):
+    """Returns a dictionary of document content (embedding, text, blob or buffer) with `modality` as its key
+    """
+    doc_content = {}
+    for chunk in doc.chunks:
+        modality = chunk.modality
+        if modality in doc_content:
+            return None
+        else:
+            doc_content[modality] = _extract_doc_content(chunk)
+    return doc_content
+
+
+class MultiModalDriver(BaseEncodeDriver):
+    """Extract multimodal embeddings from different modalities.
+    Input-Output ::
+        Input:
+        document:
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+        Output:
+        document: (embedding: multimodal encoding)
+                |- chunk: {modality: mode1}
+                |
+                |- chunk: {modality: mode2}
+
+    .. note::
+        - It traverses on the ``documents`` for which we want to apply the ``multimodal`` embedding. This way
+        we can use the `batching` capabilities for the `executor`.
+    .. warning::
+        - It assumes that every ``chunk`` of a ``document`` belongs to a different modality.
+    """
+
+    def __init__(self,
+                 traversal_paths: Tuple[str] = ('r',), *args, **kwargs):
+        super().__init__(traversal_paths=traversal_paths, *args, **kwargs)
+
+    @property
+    def position_by_modality(self):
+        if not getattr(self._exec, 'position_by_modality', None):
+            raise RuntimeError('Could not know which position of the ndarray to load to each modality')
+        return self._exec.position_by_modality
+
+    def _get_executor_input_arguments(self, content_by_modality: Dict[str, 'np.ndarray'], num_modalities: int):
+        """
+            From a dictionary ``content_by_modality`` it returns the arguments in the proper order so that they can be
+            passed to the executor.
+        """
+        input_args = [None] * num_modalities
+        for modality in self.position_by_modality.keys():
+            input_args[self.position_by_modality[modality]] = content_by_modality[modality]
```

Comment:
```suggestion
        for modality, val in self.position_by_modality.items():
            input_args[val] = content_by_modality[modality]
```

Reply:
also, isn't a List[str] enough for `position_by_modality`?

Reply:
if `position_by_modality` is a simple `List[str]` then `input_args` is simply a reorder of `content_by_modality` via:

```python
input_args = [content_by_modality[modality] for modality in position_by_modality]
```

Reply:
yes it can be done, it is more a matter of taste, I like it more explicit but we can for for this

Reply:
I went over the usage of `position_by_modality` and this really should be a list of str. This has to be fixed.

Reply:
updated